### ISSUE TITLE
Item stack details overlay stays displayed even when switching elsewhere

### ISSUE THREAD

**Issue by Nindaleth** [ja2-stracciatella/ja2-stracciatella#1777]
### What is wrong

When looking at a detail of an item stack (e.g. 4 ammo clips), clicking away the wrong way causes the stack detail overlay to stay on screen. This can be done on both the tactical and strategical views.

<details><summary>CLICK HERE to expand screenshots</summary>

This happens with ammo clips, grenades, 1st aid kits,...
![overlay-tactical](https://user-images.githubusercontent.com/6383928/216456043-63f77751-a2ad-4fe6-9db0-f27d12b59cde.png)

Persists even when switching from tactical to strategical
![overlay-strategical](https://user-images.githubusercontent.com/6383928/216456038-210069cd-0a54-4b6c-a52c-a1673fd0d142.png)

</details>

I didn't screenshot the strategical -> tactical because it's almost the same, just the overlay position differs.

| JA2S version | Game version        | Vanilla bug? | Used to work? |
| ------------ | ------------------- | ------------ | ------------- |
|  0.18.0, 0.19.0, 0.20.0, 0.21.0-dev | JA2 GOG English  |    no    |    no     |

**Reproduction**

1. right click a merc's portrait to display threir inventory
2. right click a stack of ammo
3. right click the merc's portrait again, ammo stack details overview stays on

**Env**
* Linux (Fedora 37)
* using the x86_64 AppImages

**Details**

Can be reproduced on any save, also on a new game.

Also reproduced on latest available nightly 0.21.0-20230202+fac5c69 and as early as 0.18.0 - no AppImages before that. On vanilla this doesn't happen as the mouse is not allowed to leave the stack detail overlay.

The bugged overlay also survives game loads (within one game session) and still seems to work, so it could be used to dupe stackable items that way.

Unlikely, but possibly related: #1679.

**Comment by momoko-h** [comment]
The key ring popup has the same problems, it is essentially the same code.

This is a really nice and thorough bug report. Are you working on a solution? I'm asking because sometimes people who put this much effort into a bug report become motivated to fix it, too.

**Comment by Nindaleth** [comment]
Thanks! No, despite it possibly looking that way I'm not working on a fix, I just like decent bug reports too.

**Comment by momoko-h** [comment]
IMO the root cause is Tron's refusal to implement `RestrictMouseCursor()`. As I see it, we have two options:

1. Inspect all calls of `RestrictMouseCursor()`, try to figure out what badness they try to prevent, figure out the best alternative way to prevent that badness, then write the code for it.
2. Implement `RestrictMouseCursor()`.

As you probably figured out by the way I formulated these options, I strongly lean towards option 2. It's the way the developers intended it to work and what vanilla users already know.

**Comment by Nindaleth** [comment]
After going a few years without playing JA2 at all and now being used to the stracciatella approach, restricting the area reachable by mouse feels very uncomfortable IMHO. I think there's one further solution in addition to replicating vanilla or preventing the badness directly - the developers also sometimes disable parts of the UI and their respective keyboard shortcuts. Disabling a few more UI elements while the detail overlay is displayed could also indirectly prevent this:

* disable the merc's portrait in tactical view - "map screen" button is already getting deactivated
* disable the merc's portrait and the "tactical"+"options"+"laptop" buttons in map view

However I understand I'm pretty new here without any PRs under my belt, I'll respect whichever solution you choose to fix this.

**Comment by momoko-h** [comment]
Ok, you win, this trick to add another mouse region that blocks all mouse events is already used in quite a few places, so we might as well add two more.

@Nindaleth, I guess you can build the game yourself? Would you like to give [this commit ](https://github.com/momoko-h/ja2-stracciatella/commit/1484dede89ee23cd569a1135b3da910437972ba8)a tryout before I create a pull request for it, see if you can break it somehow or if it makes anything worse?

**Comment by Nindaleth** [comment]
Sure, I can build and check custom branches/commits no problem, I'll gladly test any fix candidates if needed.

I've tested your commit and the previous obvious (well, "obvious") issue is nicely avoided. I have these findings:
- map view
  - still can switch between mercs (clicking buttons + left/right arrow) - not really a problem, only same-sector mercs are cycled through
  - still can accelerate time - most prominently, on midnight all the "subviews" reset, closing any merc details except the open the stack popup... otherwise some worse but unlíkely risks like a battle prompt showing up, hypothetically leaving the popup open either for a dead merc or after the switch to tactical (didn't get the chance to reproduce yet)
  - still can access contract details and dismiss the merc, keeping the popup open (additionally with the detail popup open their face doesn't animate when talking)
- tactical view
  - still can switch between mercs (F-keys) - not really a problem
  - stance up/down buttons in merc details are shown enabled, yet only react to keyboard shortcuts and not mouse input... this is the same as vanilla (except you can't even try to click them in vanilla)
  - sending a merc to get in a hummer and opening the stack popup before they get there also keeps the popup open when they get in
  - getting ambushed in RT by an enemy that kills the merc will keep the popup open

There's also a subtle cosmetic difference between item stack and keyring and inconsistency between tactical and map view - displaying keyring in tactical also darkens the merc portrait and stats (except the part overdrawn if game.json gui_extras == True), which doesn't happen with stacks and doesn't happen at all in map view. OTOH that's almost in line with vanilla, vanilla doesn't even darken rest of the inventory for stack popups in map view while JA2S does. I could open a new issue for this difference/inconsistency but it's such a low priority thing that I hesitate.

To sum it up a little more human readably - if the detail popup is open:
- in map view: 1) the time compression buttons and shortcuts should be disabled (which probably avoids the need to close it on midnight or incoming battle) 2) the contract button should be disabled
- in tactical view: 1) the stance up/down buttons could be clickable, talk button should be greyed out 2) the detail popup should be closed upon entering a vehicle 3) the detail popup should be closed upon dying

I didn't have the time to actually _play_ the game normally, but in the little bit of clicking around I didn't see anything getting worse.

**Comment by Nindaleth** [comment]
One more observation - when displaying details of an item ion map view, the next/prev merc buttons and shortcuts are disabled. Meanwhile in tactical switching to another merc in any way closes any opened details. That could be replicated to the stack popup too.

**Comment by selaux** [comment]
We also need to take into account that we now have touch controls, where restricting the mouse cursor probably doesn't work, as you can touch the screen anywhere.

**Comment by momoko-h** [comment]
> We also need to take into account that we now have touch controls, where restricting the mouse cursor probably doesn't work, as you can touch the screen anywhere.

Well, theoretically we could just ignore touch events outside of the restriction rectangle, but the point is moot because nobody seems to like this mechanism anyway.
> 
> Other places handle this by adding an mouse region over everything else and close if that mouse region is clicked. Would that be an option for this case?

That's almost what the commit I asked @Nindaleth to test does, except that this cover mouse region simply does not have any callbacks, so it does not react to events at all.
